src/SelfGravity/CPU_Poisson/CPU_ComputeGREP.cpp

```diff
@@ -12,189 +12,200 @@
+      double Mass_TOV_C = ( 0.5 * ( Mass_TOV[i] + Mass_TOV[i-1] ) + FourPI * CUBE( Radius[i] ) * Pres[i] / c2 )
+                        * ( 1.0 + ( Engy[i] + Pres[i] ) / ( Dens[i] * c2 ) )
+                        / SQR( 0.5 * ( Gamma_TOV[i] + Gamma_TOV[i-1] ) );
 
-      double Mass_NW_rad  = 0.5 * ( Mass_NW[i] + Mass_NW[i-1] );
-
-      double Mass_TOV_rad = ( 0.5 * ( Mass_TOV[i] + Mass_TOV[i-1] ) + FourPI * RadiusCubed * PresAve->Data[i] / c2 )
-                          * ( 1.0 + ( EngyAve->Data[i] + PresAve->Data[i] ) / ( DensAve->Data[i] * c2 ) )
-                          / SQR( 0.5 * ( Gamma_TOV[i] + Gamma_TOV[i-1] ) );
-
-      Phi_eff->Data[i] = Phi_eff->Data[i+1] - dr * NEWTON_G * ( Mass_NW_rad - Mass_TOV_rad ) / SQR( Radius[i] );
+      Phi_eff->Data[i] = Phi_eff->Data[i+1] - dr * NEWTON_G * ( Mass_NW_C - Mass_TOV_C ) / SQR( Radius[i] );
    }
 
-   double dr           = Radius[0] - Radius[1];
-   double RadiusCubed  = CUBE( Radius[0] );
-   double Mass_NW_rad  = Mass_NW[0];
-   double Mass_TOV_rad = ( Mass_TOV[0] + FourPI * RadiusCubed * PresAve->Data[0] / c2 )
-                       * ( 1.0 + ( EngyAve->Data[0] + PresAve->Data[0] ) / ( DensAve->Data[0] * c2 ) )
-                       / SQR( Gamma_TOV[0] );
+   double dr         = Radius[0] - Radius[1];
+   double Mass_NW_C  = Mass_NW[0];
+   double Mass_TOV_C = ( Mass_TOV[0] + FourPI * CUBE( Radius[0] ) * Pres[0] / c2 )
+                     * ( 1.0 + ( Engy[0] + Pres[0] ) / ( Dens[0] * c2 ) )
+                     / SQR( Gamma_TOV[0] );
 
-   Phi_eff->Data[0] = Phi_eff->Data[1] - dr * NEWTON_G * ( Mass_NW_rad - Mass_TOV_rad ) / SQR( Radius[0] ) ;
+   Phi_eff->Data[0] = Phi_eff->Data[1] - dr * NEWTON_G * ( Mass_NW_C - Mass_TOV_C ) / SQR( Radius[0] ) ;
 
 
-#ifdef GAMER_DEBUG
+// troubleshooting information
+#  ifdef GAMER_DEBUG
    if ( MPI_Rank == 0 )
    {
-      printf("\n============================================================\n");
-      printf("GREP_CENTER_METHOD : %d\n",                   GREP_CENTER_METHOD);
-      printf("Center             : %13.7e %13.7e %13.7e\n", Phi_eff->Center[0], Phi_eff->Center[1], Phi_eff->Center[2]);
-      printf("MaxRadius          : %13.7e\n",               Phi_eff->MaxRadius);
-      printf("LogBin             : %d\n",                   Phi_eff->LogBin);
-      printf("LogBinRatio        : %13.7e\n",               Phi_eff->LogBinRatio);
-      printf("Num of Iterations  : %d\n",                   GREP_MAXITER - NIter);
-      printf("NBin               : %d\n",                   NBin);
-      printf("============================================================\n");
-      printf("%5s %9s %22s %22s %22s %22s %22s %22s %22s %22s %23s\n",
-             "Bin", "NCell", "Radius", "Dens", "Engy", "Vr", "Pressure",
-             "Mass_NW", "Mass_TOV", "Gamma_TOV", "Eff_Pot");
-
+      char FileName_GREP[50];
+
+      sprintf( FileName_GREP, "GREP_Lv%02d", lv );
+      FILE *File_GREP = fopen( FileName_GREP, "w" );
```

Comment:
Yes. I think this will provide sufficient GREP profiles for debugging when the code crashes. Users can add `Step` if they want more information, however, doing so will generate too many files.

Reply:
Agree